src/libdrakvuf/os.c

```diff
@@ -128,6 +128,21 @@ bool fill_kernel_offsets(drakvuf_t drakvuf, size_t size, const char* names [][2]
     return 1;
 }
 
+bool fill_kernel_bitfields(drakvuf_t drakvuf, size_t size, const char* names [][2])
+{
+    drakvuf->bitfields = (bitfield_t)g_try_malloc0(sizeof(struct bitfield) * size );
```

Comment:
valgrind reports memory leak for this allocation for all Windows tests

Reply:
I will check. Thank You very much.

Reply:
I have checked it with address sanitizer built-in clang and fixed this and other leaks in procdump:

```
1585130254.537417 DRAKVUF v0.7-ptms.sandbox.drakvuf-0.0.0.1072-master-2-ged0c9b7 Copyright (C) 2014-2020 Tamas K Lengyel

=================================================================
==855==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 11806852 byte(s) in 39892 object(s) allocated from:
    #0 0x55c3ca7fa8ad in malloc (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x1d48ad)
    #1 0x7fa593d2442f in safe_malloc_ /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/convenience.c:110:15
    #2 0x7fa593d24576 in vmi_convert_str_encoding /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/convenience.c:165:21
    #3 0x55c3ca836b01 in drakvuf_read_unicode_common (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x210b01)
    #4 0x55c3ca837110 in drakvuf_read_unicode_va (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x211110)
    #5 0x55c3ca862a14 in win_get_process_full_name (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x23ca14)
    #6 0x55c3ca862aac in win_get_process_name (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x23caac)
    #7 0x55c3ca86b720 in win_get_process_data (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x245720)
    #8 0x55c3ca83e475 in drakvuf_get_process_data_priv (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x218475)
    #9 0x55c3ca836202 in drakvuf_get_process_data (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x210202)
    #10 0x55c3ca8b09e8 in rtlcopymemory_cb(drakvuf*, drakvuf_trap_info*) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x28a9e8)
    #11 0x55c3ca8468b7 in int3_cb (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x2208b7)
    #12 0x7fa593d13659 in process_software_breakpoint /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:687:5
    #13 0x7fa593d14d2b in process_request /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:1270:12
    #14 0x7fa593d178c9 in process_requests_4 /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:2124:15
    #15 0x7fa593d1a7c0 in xen_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:2955:11
    #16 0x7fa593d25fe2 in driver_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/driver_wrapper.h:423:12
    #17 0x7fa593d282e3 in vmi_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/events.c:1024:12
    #18 0x55c3ca850d07 in drakvuf_vmi_event_callback (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22ad07)
    #19 0x55c3ca851416 in drakvuf_poll (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22b416)
    #20 0x55c3ca85159b in drakvuf_loop (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22b59b)
    #21 0x55c3ca830c92 in drakvuf_c::loop(int) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x20ac92)
    #22 0x55c3ca82f6aa in main (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x2096aa)
    #23 0x7fa592cd7b6a in __libc_start_main /build/glibc-KRRWSm/glibc-2.29/csu/../csu/libc-start.c:308:16

Direct leak of 2011796 byte(s) in 6792 object(s) allocated from:
    #0 0x55c3ca7fa8ad in malloc (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x1d48ad)
    #1 0x7fa593d2442f in safe_malloc_ /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/convenience.c:110:15
    #2 0x7fa593d24576 in vmi_convert_str_encoding /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/convenience.c:165:21
    #3 0x55c3ca836b01 in drakvuf_read_unicode_common (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x210b01)
    #4 0x55c3ca837110 in drakvuf_read_unicode_va (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x211110)
    #5 0x55c3ca862a14 in win_get_process_full_name (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x23ca14)
    #6 0x55c3ca862aac in win_get_process_name (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x23caac)
    #7 0x55c3ca86b720 in win_get_process_data (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x245720)
    #8 0x55c3ca83e475 in drakvuf_get_process_data_priv (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x218475)
    #9 0x55c3ca836202 in drakvuf_get_process_data (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x210202)
    #10 0x55c3ca8a341c in terminate_process_cb(drakvuf*, drakvuf_trap_info*) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x27d41c)
    #11 0x55c3ca8468b7 in int3_cb (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x2208b7)
    #12 0x7fa593d13659 in process_software_breakpoint /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:687:5
    #13 0x7fa593d14d2b in process_request /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:1270:12
    #14 0x7fa593d178c9 in process_requests_4 /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:2124:15
    #15 0x7fa593d1a7c0 in xen_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:2955:11
    #16 0x7fa593d25fe2 in driver_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/driver_wrapper.h:423:12
    #17 0x7fa593d282e3 in vmi_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/events.c:1024:12
    #18 0x55c3ca850d07 in drakvuf_vmi_event_callback (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22ad07)
    #19 0x55c3ca851416 in drakvuf_poll (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22b416)
    #20 0x55c3ca85159b in drakvuf_loop (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22b59b)
    #21 0x55c3ca830c92 in drakvuf_c::loop(int) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x20ac92)
    #22 0x55c3ca82f6aa in main (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x2096aa)
    #23 0x7fa592cd7b6a in __libc_start_main /build/glibc-KRRWSm/glibc-2.29/csu/../csu/libc-start.c:308:16

Direct leak of 198 byte(s) in 1 object(s) allocated from:
    #0 0x55c3ca7fa8ad in malloc (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x1d48ad)
    #1 0x7fa593d2442f in safe_malloc_ /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/convenience.c:110:15
    #2 0x7fa593d24576 in vmi_convert_str_encoding /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/convenience.c:165:21
    #3 0x55c3ca836b01 in drakvuf_read_unicode_common (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x210b01)
    #4 0x55c3ca837110 in drakvuf_read_unicode_va (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x211110)
    #5 0x55c3ca862a14 in win_get_process_full_name (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x23ca14)
    #6 0x55c3ca862aac in win_get_process_name (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x23caac)
    #7 0x55c3ca86b720 in win_get_process_data (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x245720)
    #8 0x55c3ca83e475 in drakvuf_get_process_data_priv (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x218475)
    #9 0x55c3ca836202 in drakvuf_get_process_data (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x210202)
    #10 0x55c3ca8b27b8 in exallocatepool_cb(drakvuf*, drakvuf_trap_info*) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x28c7b8)
    #11 0x55c3ca8468b7 in int3_cb (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x2208b7)
    #12 0x7fa593d13659 in process_software_breakpoint /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:687:5
    #13 0x7fa593d14d2b in process_request /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:1270:12
    #14 0x7fa593d178c9 in process_requests_4 /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:2124:15
    #15 0x7fa593d1a7c0 in xen_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/xen/xen_events.c:2955:11
    #16 0x7fa593d25fe2 in driver_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/driver/driver_wrapper.h:423:12
    #17 0x7fa593d282e3 in vmi_events_listen /home/skovalev/src/ms/drakvuf/origin/libvmi/libvmi/events.c:1024:12
    #18 0x55c3ca850d07 in drakvuf_vmi_event_callback (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22ad07)
    #19 0x55c3ca851416 in drakvuf_poll (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22b416)
    #20 0x55c3ca85159b in drakvuf_loop (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x22b59b)
    #21 0x55c3ca830c92 in drakvuf_c::loop(int) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x20ac92)
    #22 0x55c3ca82f6aa in main (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x2096aa)
    #23 0x7fa592cd7b6a in __libc_start_main /build/glibc-KRRWSm/glibc-2.29/csu/../csu/libc-start.c:308:16

Direct leak of 168 byte(s) in 1 object(s) allocated from:
    #0 0x55c3ca7faa22 in calloc (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x1d4a22)
    #1 0x55c3ca83bf7d in fill_kernel_bitfields (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x215f7d)
    #2 0x55c3ca85c837 in set_os_windows (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x236837)
    #3 0x55c3ca832ecb in drakvuf_init (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x20cecb)
    #4 0x55c3ca830521 in drakvuf_c::drakvuf_c(char const*, char const*, char const*, output_format_t, bool, bool, bool, unsigned long) (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x20a521)
    #5 0x55c3ca82ef96 in main (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x208f96)
    #6 0x7fa592cd7b6a in __libc_start_main /build/glibc-KRRWSm/glibc-2.29/csu/../csu/libc-start.c:308:16

Direct leak of 64 byte(s) in 2 object(s) allocated from:
    #0 0x55c3ca7fabc9 in realloc (/home/skovalev/src/ms/drakvuf/origin/src/drakvuf+0x1d4bc9)
    #1 0x7fa592f8be32  (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa7e32)

SUMMARY: AddressSanitizer: 13819078 byte(s) leaked in 46688 allocation(s).
```

Reply:
Done!